File: quote/Projects/PluginModel/quote/Detail.vb
Imports System.ComponentModel
Imports System.Reflection

Imports Model

Namespace Quote

    Public Class Detail
        Inherits Common.Detail

        Private _Properties As New Common.SavableProperties

        Friend Sub New( _ 
            ByVal header As Common.Header, _
            ByVal product As Product)
            
            MyBase.New(product, "", 1)
            
            Me.QuoteHeader = header
        End Sub

        <BrowsableAttribute(False)>
        Property QuoteHeader As Header

        <BrowsableAttribute(False)>
        Public Overrides ReadOnly Property QuoteDetailProperties As Object
            Get
                Return _Properties
            End Get
        End Property

        Public Sub SetProperties(ByVal props As Common.SavableProperties)
            Me._Properties = props
        End Sub

        Public Overloads Property Qty() As Decimal
            Get
                Return MyBase.PrivateQty()
            End Get

            Set(ByVal value As Decimal)
                If Not (value = MyBase.PrivateQty()) Then
                    MyBase.SetPrivateQty(value)
                    SendEvents()
                End If
            End Set
        End Property

    End Class

End Namespace

File: quote/Projects/PluginModel/quote/Detail.cs
namespace Model.Quote
{
    using System;
    using System.ComponentModel;
    using System.Linq;

    using Model.Common;

    /// <summary>
    /// A detail object for a quote.  It can not be changed.
    /// </summary>
    public class Detail : Common.Detail
    {
        /// <summary>
        /// The properties for the quote.
        /// </summary>
        private Common.ISavableProperties properties = new QuoteSavableProperties();

        /// <summary>
        /// Initializes a new instance of the <see cref="Detail" /> class.
        /// </summary>
        /// <param name="header">The parent header.</param>
        /// <param name="product">The product.</param>
        internal Detail(Common.Header header, Product product) : base(product, string.Empty, 1)
        {
            this.QuoteHeader = header as Quote.Header;
        }

        /// <summary>
        /// Gets or sets the header for the detail.
        /// </summary>
        /// <value>The quote header.</value>
        [BrowsableAttribute(false)]
        public Header QuoteHeader { get; set; }

        /// <summary>
        /// Gets or sets the detail properties.
        /// </summary>
        /// <value>The detail properties.</value>
        [BrowsableAttribute(false)]
        public override Model.Common.ISavableProperties QuoteDetailProperties 
        {
            get { return this.properties; }
        }

        /// <summary>
        /// Gets or sets the quantity for the detail item.
        /// </summary>
        /// <value>The quantity for the detail item.</value>
        public override decimal Qty 
        {
            get 
            { 
                return this.PrivateQty(); 
            }

            set 
            {
                if (!(value == this.PrivateQty())) 
                {
                    this.SetPrivateQty(value);
                    this.OnPropertyChanged();
                }
            }
        }
        
        /// <summary>
        /// Sets the properties.
        /// </summary>
        /// <param name="props">The new properties.</param>
        public void SetProperties(Common.ISavableProperties props)
        {
            this.properties = props;
        }
    }
}